### PR TITLE
APPS: Add check for multiple 'unknown' options

### PR DESCRIPTION
**Pull request by DDvO**
* Add check for multiple 'unknown' options
* PKCS12 app: Improve readability w.r.t. `enc_flag`, renamed to `enc_name`

### CODE REVIEW

File: apps/cms.c [hunk 1]
```diff
@@ -339,6 +339,7 @@ int cms_main(int argc, char **argv)
     if (encerts == NULL || vpm == NULL)
         goto end;
 
+    opt_set_unknown_name("cipher");
```
Comment (richsalz) [inline apps/cms.c:342]: Why not make the default unknown name be `cipher` and then this improvement is much less invasive. Supporting multiple "-*" options isn't necessary.

**Comment by richsalz** [comment]
Changing opt_init is the cleaner way to do things. Yes it requires all callers to change, but at the same time it includes in one API the list of options, and which one the "wildcard" is for.  Oh well ain't gonna happen I know. And the clean foundation of the original `opt_` stuff gets more and more cruft pasted on top of it.

**Comment by DDvO** [comment]
> Changing opt_init is the cleaner way to do things.
> Yes it requires all callers to change, but at the same time it includes in one API the list of options, and which one the "wildcard" is for.

I agree.

> Oh well ain't gonna happen I know. And the clean foundation of the original `opt_` stuff gets more and more cruft pasted on top of it.

I know how if feels when others water your carefully designed API.
Yet I fear the only (efficient/feasible) way of avoiding this is to stay active yourself in its maintenance.
Sorry @richsalz that you did not succeed making me do this for you, but at some point I need to cut my activity since such things easily turn out to be a rabbit hole.

Last week I already spent several work days of my spare time to fix various OpenSSL issues, and I simply cannot afford to continue doing this.

**Comment by DDvO** [comment]
> @DDvO, thanks! Will take a look

@beldmit, have you meanwhile fixed the **external-tests**?

**Comment by beldmit** [comment]
Not yet, sorry

**Comment by beldmit** [comment]
@DDvO, no, the test is intentionally written in such a way. It tests the request with different certificates each of them has its own hash.

**Comment by DDvO** [comment]
@beldmit, this PR still fails due to your external tests.
When are you going to fix them?

**Comment by beldmit** [comment]
@DDvO no, the tests test exactly what is intended - and your changes break this behavior, as I denoted on https://github.com/openssl/openssl/pull/16416#issuecomment-918461776

**Comment by beldmit** [comment]
The tests tests the behavior "specify a hash for each certificate". Your changes seem to break it.

**Comment by DDvO** [comment]
Thanks @beldmit - now I got it!

**Comment by beldmit** [comment]
> Note the weird argument -[hash_short_name [alg_hash $alg]]
> which is subsequently overridden by -sha1 such that so far it went unnoticed.

It shouldn't be an overwrite, AFAIK

**Comment by DDvO** [comment]
> > Note the weird argument -[hash_short_name [alg_hash $alg]]
> > which is subsequently overridden by -sha1 such that so far it went unnoticed.
> 
> It shouldn't be an overwrite, AFAIK

Well, I meant that `-[hash_short_name [alg_hash $alg]]` gets replaced/instantated by `-sha1`.
Would be nice if the test output reflected this, i.e., did now show
```
openssl "ocsp $client_args -[hash_short_name [alg_hash $alg]] \
    -cert U_ocsp_$alg_fn/cert.pem -sha1 -cert U_ocsp2_$alg_fn/cert.pem -reqout request5.der"
```
but
```
openssl "ocsp $client_args -sha1 \
    -cert U_ocsp_$alg_fn/cert.pem -sha1 -cert U_ocsp2_$alg_fn/cert.pem -reqout request5.der"
```
where it would be better if also `$client` and `$alg_fn` were not shown as variable names, but their actual values.
Then it would be simpler to understand and maybe even reproduce the test case.

**Comment by DDvO** [comment]
Yet meanwhile I noticed that the full instantiation of the command line already is shown:
```
/home/runner/work/openssl/openssl/apps/openssl ocsp -issuer ocspCA-gost2012_512_B/cacert.pem -CAfile ocspCA-gost2012_512_B/cacert.pem 
  -md_gost12_512 -cert U_ocsp_gost2012_512_B/cert.pem 
  -sha1 -cert U_ocsp2_gost2012_512_B/cert.pem -reqout request5.der
```
before the partly instantiated command line is shown:
```
openssl "ocsp $client_args 
  -[hash_short_name [alg_hash $alg]] -cert U_ocsp_$alg_fn/cert.pem
  -sha1 -cert U_ocsp2_$alg_fn/cert.pem -reqout request5.der" 
```

**Comment by beldmit** [comment]
@DDvO great!

**Comment by DDvO** [comment]
I've also made clear now in the doc and help output that the `-cert` and `-serial` options can be given multiple times (while for the digest name options this was already mentioned).

**Comment by DDvO** [comment]
@paulidale can you please re-approve,
now that the external-tests issue is solved?

**Comment by DDvO** [comment]
Ping for re-review

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by DDvO** [comment]
Merged - thanks @beldmit for the approval

**Pull request closed without merging**